src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +404,202 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+        device_src, src->dictionary, device_dst, dst->dictionary, stream,
+        unknown_buffer_size_count));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaCreateStream(struct ArrowDevice* src,
+                                                  struct ArrowDevice* dst, CUstream* out,
+                                                  struct ArrowError* error) {
+  struct ArrowDeviceCudaPrivate* private_data;
+
+  if (dst->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)dst->private_data;
+  } else if (src->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)src->private_data;
+  } else {
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuCtxPushCurrent(private_data->cu_context),
+                               "cuCtxPushCurrent", NULL);
+  CUresult err = cuStreamCreate(out, CU_STREAM_NON_BLOCKING);
+  CUcontext unused;
+  NANOARROW_CUDA_ASSERT_OK(cuCtxPopCurrent(&unused));
+  NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamCreate", error);
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaArrayViewCopyInternal(
+    struct ArrowDevice* device_src, struct ArrowDeviceArrayView* src,
+    struct ArrowDevice* device_dst, struct ArrowArray* dst, CUstream* stream,
+    struct ArrowError* error) {
+  CUresult err;
+  int code;
+  int64_t unknown_buffer_size_count = 0;
+
+  NANOARROW_RETURN_NOT_OK_WITH_ERROR(
+      ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst, dst,
+                                          stream, &unknown_buffer_size_count),
+      error);
+
+  if (unknown_buffer_size_count > 0) {
+    // If buffers with unknown size were encountered, we need to copy all of their
+    // sizes to the CPU, synchronize, then copy all of the buffers once the CPU
+    // knows how many bytes from each buffer need to be copied.
+
+    // Issue all required async copies
+    NANOARROW_RETURN_NOT_OK_WITH_ERROR(
+        ArrowDeviceCudaResolveBufferSizesAsync(device_src, &src->array_view, stream),
+        error);
+
+    // Wait for them all to complete
+    NANOARROW_CUDA_RETURN_NOT_OK(cuStreamSynchronize(*stream), "cuStreamSynchronize",
+                                 error);
+
+    // Issue async copies for data buffers. This will skip buffers that are already
+    // being copied.
+    unknown_buffer_size_count = 0;
+    NANOARROW_RETURN_NOT_OK_WITH_ERROR(
+        ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst, dst,
+                                            stream, &unknown_buffer_size_count),
+        error);
+
+    // There should not have been any buffers with unknown size in this round of copying
+    NANOARROW_DCHECK(unknown_buffer_size_count == 0);
+  }
+
+  // TODO: Populate the sync event instead of synchornizing
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuStreamSynchronize(*stream), "cuStreamSynchronize",
+                               error);
```

Comment:
As long as you call `cuStreamWaitEvent` you don't have to manually synchronize the stream, you can leave that to the caller/consumer. We don't need the CPU to synchronize and wait, we just need the stream to wait for the event before it gets there.

Reply:
If we're not populating a sync event then if call `cuStreamWaitEvent` with the given stream and someone tries to touch `dst` from a different stream we end up with a race. We either need to synchronize the stream or populate the sync event.